migrations/2023-08-31-205559_add_image_upload/up.sql

```diff
@@ -0,0 +1,12 @@
+CREATE TABLE image_upload (
+    id serial PRIMARY KEY,
+    local_user_id int REFERENCES local_user ON UPDATE CASCADE ON DELETE CASCADE NOT NULL,
+    pictrs_alias text NOT NULL,
+    pictrs_delete_token text NOT NULL,
+    published timestamptz DEFAULT now() NOT NULL
+);
```

Comment:
Nice, we should have added this a while ago. 

@asonix Is there any way to ensure that this is the **first** / original uploader, rather than using the alias? It'd be nice if this table didn't store the aliases, but pictrs's official *unique* ID. 

Then deletes can delete the image, not just an alias.

Reply:
We actually do want to delete just the alias, especially if you are doing something like purgeuser, because we don't want to delete everything they have uploaded, which may point to images that others have uploaded.

Reply:
makotech is right. You do want the alias here & not the internal unique ID. pict-rs intentionally does not expose a way to retrieve the internal ID via the alias. Plus, aliases are already unique. 

When a user deletes a post containing an image, that `alias` should be deleted, but the the image _should not_ be purged. This would impact availability of that image in other posts that are not deleted.

Purge exists as an internal endpoint for use when an admin decides an image absolutely needs to be removed from all posts at the cost of breaking image links.

Reply:
There are a few use cases here tho if we allow a ton of duplicate images / aliases in the DB, counted as local_user uploads:

- Malicious images. If we delete only the alias link, then that image could still exist on the filesystem, if other people (or the same malicious actor) uploaded it several times as several different aliases. Because you have a ton of duplicate images, you don't know who the original uploader actually is, and whether to purge or just remove the alias.
- Non-malicious images. Having a ton of duplicate images linked to local_users means we don't know whether to purge or delete the alias again, because purging a user could remove safe images added by other people.

Shouldn't the point of this table, be to show who the original uploader is, not someone who possibly uploaded a duplicate?

Reply:
The internal purge endpoint will purge an image given any alias. There's no distinction in pict-rs itself between an "original" upload and a "duplicate" upload. They are both considered aliases of the same hash.

Caring about who is the original uploader vs who uploaded a duplicate doesn't make sense to me. Ultimately, if duplicates exist, it just means multiple people uploaded the same file.

Reply:
if you're concerned about having dangling aliases attached to upload records after purging an image, you can use pict-rs' internal aliases endpoint to retrieve all aliases for a hash before purging it.

Reply:
> Caring about who is the original uploader vs who uploaded a duplicate doesn't make sense to me. Ultimately, if duplicates exist, it just means multiple people uploaded the same file.

Because the main issue, that's the reason for this iirc, is purging users. Lets say someone creates an account and : 

- Uploads a few *original* malicious images.
- Uploads a few non-malicous copies of other peoples (aliases)

When we do a user purge, how do we know we aren't deleting non-malicious images that don't belong to them?

Reply:
Not really. The point of this PR is to really log who is uploading what, and to provide a list of images to delete when purging a user. 

Pictrs creates an alias for an image if the image sha256 matches, so as to prevent duplicates. If only one alias exists for an image, then deleting the alias will also delete the image.

There is a separate user story that can be made for actually purging an image (original image and all aliases), but we currently don't have that functionality in lemmy. For every other use case though, we don't want to purge images just purely based on the uploader, since they could be uploading normal valid images which might be duplicates of other valid images.

Reply:
I agree with makotech, if an admin is deleting a user, that user's images should be `deleted` and not `purged`. In the majority of cases this will result in the media being removed.

If another user has also uploaded the Bad Image then that other user can also be deleted, which will result in the image being deleted.

Reply:
That would cause problems because another user might have uploaded the same image, so that would still be used. In fact this could be abused by a malicious user if he reuploads something like the site icon, then does something bad to get purged. In that case the site icon would get purged as well and its not what we want.